### PR TITLE
Fix #1712 - remove reflection from validation code in MutableObjectModel...

### LINKED ISSUE #1712: MutableObjectModelBinder.GetPropertyValidationInfo should not reflect over `ModelType.Properties`
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs#L305-L306

In the MutableObjectModelBinder we attempt to look for `BindingBehaviorAttribute` on properties. As part of this, we iterate over all properties on the `ModelType`. This is problematic in

1. There is a mismatch in the properties we discover as part of  `ModelMetadata` and the ones that are   discovered here. For instance `ModelMetadata` ignores indexers versus this does not. Consequently a model with indexers causes an exception to be thrown at https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs#L309.

2. Reflecting on the type as opposed to using property metadata seems to be against the established pattern of reading cached metadata.

### PR DESCRIPTION
**Pull request by rynowak**
...Binder

This change moves [BindingBehavior(...)] and friends into the model
metadata layer, and removes the reflection code from
MutableObjectModelBinder that was looking for them previously.

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs [hunk 1]
```diff
@@ -334,29 +337,6 @@ internal static PropertyValidationInfo GetPropertyValidationInfo(ModelBindingCon
                 if (requiredValidator != null)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs]: just to be clear: this change leaves `[Required]` also implicitly meaning `BindingBehavior.Required`?

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/BindingMetadata.cs [hunk 2]
```diff
@@ -29,6 +29,13 @@ public class BindingMetadata
         public Type BinderType { get; set; }
 
         /// <summary>
+        /// Gets or sets a value indicating whether or not the property can be model bound.
+        /// Will be ignored if the model metadata being created is not a property.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/BindingMetadata.cs]: nit: sentence doesn't scan.  think you mean `Will be ignored if the current <see cref="ModelMetadata"/> is not for a property.` or perhaps `Will be ignored unless creating a model for a property.`

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelMetadata.cs [hunk 3]
```diff
@@ -167,7 +167,17 @@ public string PropertyName
...
+        /// applicable when the current instance represents a property.
+        /// </summary>
+        /// <remarks>
+        /// If <c>true</c> then the model value is considered supported by model-binding and can be set
+        /// based on provided input in the request.
+        /// </remarks>
+        public abstract bool IsBindingAllowed { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether or not the model value is required by model-binding. This is only
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelMetadata.cs]: the "model binding" to "model-binding" switch doesn't make sense to me.  see a few mentions of "model-binding" in existing doc comments but far more using "model binding".  the majority is correct in this case.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 4]
```diff
@@ -256,11 +256,34 @@ public override bool HtmlEncode
         }
 
         /// <inheritdoc />
+        public override bool IsBindingAllowed
+        {
+            get
+            {
+                if (MetadataKind == ModelMetadataKind.Property)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs]: these checks look redundant and cause the property to have defaults both here and in the `BindingMetadata` class.  is this done to avoid filling in the `BindingMetadata` property until and unless it's absolutely required?

**Comment by dougbu** [comment]
:watch: and haven't looked in detail at the tests yet

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs]: It's defense in depth. I fixed untracked bugs in this exact logic for a similar property in the last PR. It's far easier to just write the correct code in both places.

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelMetadata.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelMetadata.cs]: sure. I think I felt like using *model-binding* one day and started doing it :+1: I'll clean it up.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs]: Yes. I logged a separate bug for that item. https://github.com/aspnet/Mvc/issues/2407 at this point it's getting hard to track what the impact of #1712 is supposed to be, so it felt like another issue was warranted.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs [hunk 5]
```diff
@@ -712,7 +712,7 @@ public void PropertySetter_Null_ForPrivateSetProperty()
             Assert.Null(metadata.PropertySetter);
             Assert.NotNull(metadata.PropertyGetter);
         }
-
+
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: nit remove.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs (hunk 1 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs]: to be clear: the issue mentioned in #2407 was there even before this PR.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs [hunk 6]
```diff
@@ -147,5 +147,358 @@ public void GetBindingDetails_FindsBindingSource_IfNullFallsBack()
...
+            Assert.True(context.BindingMetadata.IsBindingRequired);
+        }
+
+        [Fact]
+        public void GetBindingDetails_FindsBindRequired_OnProperty()
+        {
+            // Arrange
+            var propertyAttributes = new object[]
+            {
+                new BindingBehaviorAttribute(BindingBehavior.Required),
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: Shouldnt this be 
`new BindRequiredAttribute();`

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultBindingMetadataProvider.cs [hunk 7]
```diff
@@ -53,6 +54,21 @@ public void GetBindingMetadata([NotNull] BindingMetadataProviderContext context)
...
                     predicateProviders);
             }
+
+            if (context.Key.MetadataKind == ModelMetadataKind.Property)
+            {
+                // BindingBehavior can fall back to attributes on the Container Type, but we should ignore
+                // attributes on the Property Type.
+                var bindingBehavior =
+                    context.PropertyAttributes.OfType<BindingBehaviorAttribute>().FirstOrDefault() ??
+                    context.Key.ContainerType.GetTypeInfo().GetCustomAttribute<BindingBehaviorAttribute>(inherit: true);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultBindingMetadataProvider.cs]: Isn't this an additional behavior we are defining?
Can we do it separately? 
That will help us consider the alternatives like
Shouldn't the attribute on the type containing the property apply to whether a model of that type gets model bound or not?

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs [hunk 8]
```diff
@@ -147,5 +147,358 @@ public void GetBindingDetails_FindsBindingSource_IfNullFallsBack()
...
+            // Assert
+            Assert.True(context.BindingMetadata.IsBindingAllowed);
+            Assert.True(context.BindingMetadata.IsBindingRequired);
+        }
+
+        [Fact]
+        public void GetBindingDetails_FindsBindRequired_OnContainerClass()
+        {
+            // Arrange
+            var propertyAttributes = new object[]
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: nit: you could inline this as well

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs [hunk 9]
```diff
@@ -147,5 +147,358 @@ public void GetBindingDetails_FindsBindingSource_IfNullFallsBack()
...
+            {
+            };
+
+            var context = new BindingMetadataProviderContext(
+                ModelMetadataIdentity.ForProperty(typeof(string), "Property", typeof(BindRequiredOnClass)),
+                new ModelAttributes(propertyAttributes, typeAttributes: new object[0]));
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: does this pass? you are passing in an empty typeAttribute array?

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs [hunk 10]
```diff
@@ -147,5 +147,358 @@ public void GetBindingDetails_FindsBindingSource_IfNullFallsBack()
...
+        public void GetBindingDetails_FindsBindNever_OnContainerClass()
+        {
+            // Arrange
+            var propertyAttributes = new object[]
+            {
+            };
+
+            var context = new BindingMetadataProviderContext(
+                ModelMetadataIdentity.ForProperty(typeof(string), "Property", typeof(BindNeverOnClass)),
+                new ModelAttributes(propertyAttributes, typeAttributes: new object[0]));
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: does this pass?

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultBindingMetadataProvider.cs (hunk 7 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultBindingMetadataProvider.cs]: @harshgMSFT this is the current behavior of this attribute, nothing has changed. I added functional tests for this in a previous PR.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs (hunk 9 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: Yes. Please read the comments in the relevant code, BindingBehavior looks at the container type not the property type.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultBindingMetadataProvider.cs (hunk 7 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultBindingMetadataProvider.cs]: sorry totally missed https://github.com/aspnet/Mvc/pull/2413/files#diff-62a5557f611407fe0ee3bd9e9e2a1393L340

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs]: Yes.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/BindingMetadata.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/BindingMetadata.cs]: changed it to `Will be ignored if the model metadata being created does not represent a property`

**Comment by harshgMSFT** [comment]
:shipit: after fixing minor issue in test.

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelMetadata.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelMetadata.cs]: changed all occurrences in this class to `model binding`

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs (hunk 6 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: :+1:

**Comment by rynowak** [comment]
@dougbu @harshgMSFT updated

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs (hunk 6 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: isn't `[BindRequired]` tested in the next test?  this one is about `BindingBehavior.Required`

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs [hunk 11]
```diff
@@ -147,5 +147,350 @@ public void GetBindingDetails_FindsBindingSource_IfNullFallsBack()
...
+            Assert.True(context.BindingMetadata.IsBindingRequired);
+        }
+
+        [Fact]
+        public void GetBindingDetails_FindsBindRequired_OnContainerClass()
+        {
+            // Arrange
+            var context = new BindingMetadataProviderContext(
+                ModelMetadataIdentity.ForProperty(typeof(string), "Property", typeof(BindRequiredOnClass)),
+                new ModelAttributes(propertyAttributes: new object[0], typeAttributes: new object[0]));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: hmm, should `ModelAttributes` contain the attributes on the container class as well?  this test makes it very clear that there's a third category of attributes and that `ModelAttributes` ignores this category.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs (hunk 6 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: Yep, if you look at the tests **now** they do the right thing. I'm not sure why github didn't hide the original comment, because I did make a change here.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs [hunk 12]
```diff
@@ -147,5 +147,350 @@ public void GetBindingDetails_FindsBindingSource_IfNullFallsBack()
...
+            // Act
+            provider.GetBindingMetadata(context);
+
+            // Assert
+            Assert.False(context.BindingMetadata.IsBindingAllowed);
+            Assert.False(context.BindingMetadata.IsBindingRequired);
+        }
+
+        [Fact]
+        public void GetBindingDetails_OverrideBehaviorOnClass_OverrideWithOptional()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: nit `..._BehaviorOnClass_OverriddenWithOptional` and similar for a couple of tests below.  hard to parse name and understand what is overriding what.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs [hunk 13]
```diff
@@ -147,5 +147,350 @@ public void GetBindingDetails_FindsBindingSource_IfNullFallsBack()
...
+        public void GetBindingDetails_OverrideInheritedBehaviorOnClass_OverrideWithRequired()
+        {
+            // Arrange
+            var propertyAttributes = new object[]
+            {
+                new BindRequiredAttribute()
+            };
+
+            var context = new BindingMetadataProviderContext(
+                ModelMetadataIdentity.ForProperty(typeof(string), "Property", typeof(InheritedBindNeverOnClass)),
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: would like to see another test using `InheritedBindNeverOnClass` but without an override on the property

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs [hunk 14]
```diff
@@ -147,5 +147,350 @@ public void GetBindingDetails_FindsBindingSource_IfNullFallsBack()
...
+
+        private class InheritedBindNeverOnClass : BindNeverOnClass
+        {
+        }
+
+        [BindRequired]
+        private class BindRequiredOnClass
+        {
+            public string Property { get; set; }
+        }
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: aren't testing precedence of multiple attributes in class hierarchy. would be good to have a test or two using a class like
``` c#
[BindNever]
class BindNeverOnClassTotallyWins : BindRequiredOnClass
{
}
```

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs [hunk 15]
```diff
@@ -869,6 +948,21 @@ private class RequiredMember
...
+            [BindRequired]
+            public int BindRequiredOverride { get; set; }
+
+            [BindingBehavior(BindingBehavior.Optional)]
+            public int BindingBehaviorOptionalOverride { get; set; }
+
+            [BindingBehavior(BindingBehavior.Never)]
+            public int BindingBehaviorNever { get; set; }
+
+            public int BindingBehaviorNeverSetOnClass { get; set; }
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: `BindingBehaviorInheritedFromClass`

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs [hunk 16]
```diff
@@ -869,6 +948,21 @@ private class RequiredMember
...
+        private class BindingBehaviorModel
+        {
+            [BindRequired]
+            public int BindRequiredOverride { get; set; }
+
+            [BindingBehavior(BindingBehavior.Optional)]
+            public int BindingBehaviorOptionalOverride { get; set; }
+
+            [BindingBehavior(BindingBehavior.Never)]
+            public int BindingBehaviorNever { get; set; }
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: `BindingBehaviorDuplicatedFromClass`

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: or `BindingBehaviorNeverDuplicatedFromClass`

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs [hunk 17]
```diff
@@ -726,6 +726,85 @@ public void Metadata_Null_ForPrivateGetProperty()
...
+            // Act
+            var metadata = metadataProvider.GetMetadataForProperty(type, propertyName);
+
+            // Assert
+            Assert.True(metadata.IsBindingAllowed);
+            Assert.True(metadata.IsBindingRequired);
+        }
+
+        [Fact]
+        public void BindingBehavior_Override_OptionalOnProperty()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: nit: prefer naming scheme used below to what you're using for these two `_Override_` tests.  for this one how 'bout `BindingBehavior_Optional_SetOnPropertyOverridingClass`?

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs]: :ok: though I remain a bit concerned about providing default values twice.

**Comment by dougbu** [comment]
:watch: mostly for test gaps and potential `ModelAttributes` extension

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs (hunk 16 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: sure.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs (hunk 11 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: I don't really like the idea because it's only used in a 2 very specialized places.  If we end up doing https://github.com/aspnet/Mvc/issues/2426 then I think we should do it.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs (hunk 12 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: this just seems different not better, don't care

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs (hunk 13 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: :+1:

Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: Added `GetBindingDetails_FindsBindNever_OnBaseClass`

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs (hunk 11 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: :ok: fine as-is then

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs (hunk 14 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultBindingMetadataProviderTest.cs]: Sure, added `GetBindingDetails_OverrideBehaviorOnBaseClass_OverrideWithRequired_OnClass`

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs (hunk 15 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: I'm avoiding using the term inherited for anything other than class -> class inheritance.

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: inferred? FallsBackTo? :smiley_cat:

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs (hunk 17 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: don't see the improvement sorry

**Comment by rynowak** [comment]
@dougbu updated - added tests

**Comment by dougbu** [comment]
:shipit:
